### PR TITLE
MD Algorithm to convert from Q to HKL

### PR DESCRIPTION
**Pull request by ckendrick**
**Description of work.**

Creates a new algorithm `ConvertQtoHKLMDHisto` which converts a `MDEventWorkspace` in QSample to a `MDHistoWorkspace` in HKL. The conversion can use the UB matrix from the input workspace, or from a `PeaksWorkspace`. 

**To test:**

Verify the unit test `ConvertQtoHKLMDHistoTest` passes.

The following script can be used to make sure that the conversion to HKL looks correct and that peaks are overlaid properly.
```
data = HB3AAdjustSampleNorm(Filename="/HFIR/HB3A/IPTS-19962/shared/autoreduce/HB3A_exp0737_scan0078.nxs")

peaks = HB3AFindPeaks(InputWorkspace=data, 
                      CellType="Tetragonal",
                      Centering="I",
                      PeakDistanceThreshold=0.25, 
                      DensityThresholdFactor=5000,
                      Wavelength=1.551)

TransformHKL(PeaksWorkspace=peaks, HKLTransform="-0.3333333,0,0,0,0.3333333,0,0,0,-1")

mdhist = ConvertQtoHKLMDHisto(InputWorkspace=data,
                              PeaksWorkspace=peaks)
```
When plotting in SliceViewer and overlaying the peaks, the peaks should appear correctly over the data:

scd141_scan78_hkl_peaks

*There is no associated issue.*

The version of this into `ornl-next` is #30136 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/ConvertQtoHKLMDHisto.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+        return issues
+
+    def PyExec(self):
+        input_ws = self.getProperty("InputWorkspace").value
+
+        extents = self.getProperty("Extents").value
+        bins = self.getProperty("Bins").value
+
+        if self.getProperty("FindUBFromPeaks").value:
```
Comment (rosswhitfield) [inline Framework/PythonInterface/plugins/algorithms/ConvertQtoHKLMDHisto.py:127]: You could simplfy this by removing the `FindUBFromPeaks` option and just have it that if a `PeaksWorkspace` is provided use the UB from it.

**Review by rosswhitfield — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/plugins/algorithms/ConvertQtoHKLMDHisto.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+class ConvertQtoHKLMDHisto(PythonAlgorithm):
+
+    _lattice = None
+
+    def category(self):
+        return "MDAlgorithms\\Creation"
+
+    def seeAlso(self):
+        return ["ConvertWANDSCDtoQ", "BinMD"]
```
Comment (rosswhitfield) [inline Framework/PythonInterface/plugins/algorithms/ConvertQtoHKLMDHisto.py:23]: Should add MDNorm

File: docs/source/algorithms/ConvertQtoHKLMDHisto-v1.rst [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+.. properties::
+
+Description
+-----------
+
+This algorithm provides a way to convert from a `MDEventWorkspace` in QSample units to a `MDHistoWorkspace` in HKL
+units. By default, the UB matrix from the `InputWorkspace` is used unless `FindUBFromPeaks` is true. In this case, the
+UB matrix from the specified `PeaksWorkspace` will be used. An optional transformation can be applied to the peak
+UB matrix to better align peaks in HKL space (see :ref:`TransformHKL <algm-TransformHKL>`).
+
```
Comment (rosswhitfield) [inline docs/source/algorithms/ConvertQtoHKLMDHisto-v1.rst:16]: Can you add a note that "this doesn't correctly take into account the normalization of the output, it's simply binning the data in HKL. Look at MDNorm (it using TOF) to correctly take into account the normalization when converting to HKL"

**Review by rosswhitfield — CHANGES_REQUESTED** [review]

File: docs/source/algorithms/ConvertQtoHKLMDHisto-v1.rst (hunk 3 above)
Comment (AndreiSavici) [reply docs/source/algorithms/ConvertQtoHKLMDHisto-v1.rst:16]: You should add some formula, or link to the lattice concept as well

**Review by rosswhitfield — APPROVED** [review]

**Pull request merged**